orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h

```diff
@@ -0,0 +1,180 @@
+  * begin or resume training from a given checkpoint state for the given onnx models.
+  *
+  * \param[in] env Environment to be used for the training session.
+  * \param[in] options Session options that the user can customize for this training session.
+  * \param[in] checkpoint_state Training states that the training session uses as a starting point for training.
+  * \param[in] train_model_path Model to be used to perform training that can be generated using the offline tooling library.
+  * \param[in] eval_model_path Model to be used to perform evaluation that can be generated using the offline tooling library.
+  * \param[in] optimizer_model_path Model to be used to the optimizer step for weight updates. The model can be generated using the offline tooling library.
+  * \param[out] out Created training session.
+  *
+  * \snippet{doc} snippets.dox OrtStatus Return Value
+  *
+  */
+  ORT_API2_STATUS(CreateTrainingSession, _In_ const OrtEnv* env, _In_ const OrtSessionOptions* options,
+                  _Inout_ OrtCheckpointState* checkpoint_state, _In_ const ORTCHAR_T* train_model_path,
+                  _In_ const ORTCHAR_T* eval_model_path, _In_ const ORTCHAR_T* optimizer_model_path,
+                  _Outptr_ OrtTrainingSession** out);
+
+  /** \brief Retrieves the number of user outputs in the training model.
+  *
+  * This function returns the number of outputs of the training model so that the user can
+  * allocate space for the number of outputs when TrainStep is invoked.
+  *
+  * \param[in] sess The training session which has working knowledge of the training model.
+  * \param[out] out Number of user outputs in the training model.
+  *
+  * \snippet{doc} snippets.dox OrtStatus Return Value
+  *
+  */
+  ORT_API2_STATUS(TrainingSessionGetTrainModeOutputCount, _In_ const OrtTrainingSession* sess, _Out_ size_t* out);
+
+  /** \brief Retrieves the number of user outputs in the eval model.
+  *
+  * This function returns the number of outputs of the eval model so that the user can
+  * allocate space for the number of outputs when EvalStep is invoked.
+  *
+  * \param[in] sess The training session which has working knowledge of the eval model.
+  * \param[out] out Number of user outputs in the eval model.
+  *
+  * \snippet{doc} snippets.dox OrtStatus Return Value
+  *
+  */
+  ORT_API2_STATUS(TrainingSessionGetEvalModeOutputCount, _In_ const OrtTrainingSession* sess, _Out_ size_t* out);
+
+  /** \brief Reset the training model gradients to zero lazily.
+  *
+  * This function sets the internal state of the training session such that the training model gradients
+  * will be reset just before the new gradients are computed on the next invocation of TrainStep.
+  *
+  * \param[in] sess The training session which has working knowledge of the eval model.
+  *
+  * \snippet{doc} snippets.dox OrtStatus Return Value
+  *
+  */
+  ORT_API2_STATUS(ResetGrad, _In_ const OrtTrainingSession* session);
```

Comment:
perhaps ResetGrad's OrtTrainingSession should be non-const?
also I noticed that the ORT C API Run function takes a non-const OrtSession. should we do something similar for the Step functions?

Reply:
oh, I see that TrainingSession implements these in const methods by calling other methods through a pointer. despite what c++ allows, should those be const methods if they are actually calling non-const methods on an owned object?

Reply:
you're right. I think we should preserve the meaning of constness here. I will update the methods so they are non const if they update any member.